sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,432 @@
+      {PADCTRL_DIO_PADS_REG_OFFSET, PADCTRL_DIO_PADS_ATTR1_MASK,
+       PADCTRL_DIO_PADS_ATTR1_OFFSET, 1},
+      {PADCTRL_DIO_PADS_REG_OFFSET, PADCTRL_DIO_PADS_ATTR2_MASK,
+       PADCTRL_DIO_PADS_ATTR2_OFFSET, 2},
+      {PADCTRL_DIO_PADS_REG_OFFSET, PADCTRL_DIO_PADS_ATTR3_MASK,
+       PADCTRL_DIO_PADS_ATTR3_OFFSET, 3},
+  };
+
+  std::vector<AttrField> mio_attr_fields_ = {
+      {PADCTRL_MIO_PADS0_REG_OFFSET, PADCTRL_MIO_PADS0_ATTR0_MASK,
+       PADCTRL_MIO_PADS0_ATTR0_OFFSET, 0},
+      {PADCTRL_MIO_PADS0_REG_OFFSET, PADCTRL_MIO_PADS0_ATTR1_MASK,
+       PADCTRL_MIO_PADS0_ATTR1_OFFSET, 1},
+      {PADCTRL_MIO_PADS0_REG_OFFSET, PADCTRL_MIO_PADS0_ATTR2_MASK,
+       PADCTRL_MIO_PADS0_ATTR2_OFFSET, 2},
+      {PADCTRL_MIO_PADS0_REG_OFFSET, PADCTRL_MIO_PADS0_ATTR3_MASK,
+       PADCTRL_MIO_PADS0_ATTR3_OFFSET, 3},
+      {PADCTRL_MIO_PADS1_REG_OFFSET, PADCTRL_MIO_PADS1_ATTR4_MASK,
+       PADCTRL_MIO_PADS1_ATTR4_OFFSET, 4},
+      {PADCTRL_MIO_PADS1_REG_OFFSET, PADCTRL_MIO_PADS1_ATTR5_MASK,
+       PADCTRL_MIO_PADS1_ATTR5_OFFSET, 5},
+      {PADCTRL_MIO_PADS1_REG_OFFSET, PADCTRL_MIO_PADS1_ATTR6_MASK,
+       PADCTRL_MIO_PADS1_ATTR6_OFFSET, 6},
+      {PADCTRL_MIO_PADS1_REG_OFFSET, PADCTRL_MIO_PADS1_ATTR7_MASK,
+       PADCTRL_MIO_PADS1_ATTR7_OFFSET, 7},
+      {PADCTRL_MIO_PADS2_REG_OFFSET, PADCTRL_MIO_PADS2_ATTR8_MASK,
+       PADCTRL_MIO_PADS2_ATTR8_OFFSET, 8},
+      {PADCTRL_MIO_PADS2_REG_OFFSET, PADCTRL_MIO_PADS2_ATTR9_MASK,
+       PADCTRL_MIO_PADS2_ATTR9_OFFSET, 9},
+      {PADCTRL_MIO_PADS2_REG_OFFSET, PADCTRL_MIO_PADS2_ATTR10_MASK,
+       PADCTRL_MIO_PADS2_ATTR10_OFFSET, 10},
+      {PADCTRL_MIO_PADS2_REG_OFFSET, PADCTRL_MIO_PADS2_ATTR11_MASK,
+       PADCTRL_MIO_PADS2_ATTR11_OFFSET, 11},
+      {PADCTRL_MIO_PADS3_REG_OFFSET, PADCTRL_MIO_PADS3_ATTR12_MASK,
+       PADCTRL_MIO_PADS3_ATTR12_OFFSET, 12},
+      {PADCTRL_MIO_PADS3_REG_OFFSET, PADCTRL_MIO_PADS3_ATTR13_MASK,
+       PADCTRL_MIO_PADS3_ATTR13_OFFSET, 13},
+      {PADCTRL_MIO_PADS3_REG_OFFSET, PADCTRL_MIO_PADS3_ATTR14_MASK,
+       PADCTRL_MIO_PADS3_ATTR14_OFFSET, 14},
+      {PADCTRL_MIO_PADS3_REG_OFFSET, PADCTRL_MIO_PADS3_ATTR15_MASK,
+       PADCTRL_MIO_PADS3_ATTR15_OFFSET, 15}};
+
+  // Vector of all attributes, initialise with the basic attributes.
+  std::vector<dif_padctrl_attr_t> attributes_ = {
+      kDifPadctrlAttrIOInvert, kDifPadctrlAttrOpenDrain,
+      kDifPadctrlAttrPullDown, kDifPadctrlAttrPullUp,
+      kDifPadctrlAttrKeeper,   kDifPadctrlAttrWeakDrive};
+
+  void SetUp() override {
+    // Add additional attributes to attributes_ vector during setup.
+    for (dif_padctrl_attr_t attr = kDifPadctrlAttrAdditionalFirst;
+         attr <= kDifPadctrlAttrLast; attr = (dif_padctrl_attr_t)(attr + 1)) {
+      attributes_.push_back(attr);
+    }
+  }
+
+  void DeviceInit() {
+    EXPECT_READ32(PADCTRL_REGEN_REG_OFFSET, 1);
+
+    // Extract unique register offsets from dio_attr_fields_
+    std::set<ptrdiff_t> dio_pad_reg_offsets;
```

Comment:
Since you're using the fact that `std::set` is ordered, you might want to note this in a comment. It's pretty common for folks to use `std::set` when they meant `std::unordered_set`, so people might think this is a mistake when it's actually intended.